cluster.go

```diff
@@ -0,0 +1,464 @@
+package redis
+
+import (
+	"container/list"
+	"errors"
+	"io"
+	"math/rand"
+	"net"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type ClusterClient struct {
+	commandable
+
+	addrs []string
+	slots [][]string
+	conns *lruPool
+	opt   *ClusterOptions
+
+	cachemx sync.RWMutex
+	_reload uint32
+}
+
+func NewClusterClient(opt *ClusterOptions) (*ClusterClient, error) {
+	addrs, err := opt.getAddrs()
+	if err != nil {
+		return nil, err
+	}
+
+	client := &ClusterClient{
+		addrs:   addrs,
+		conns:   newLRUPool(opt),
+		opt:     opt,
+		_reload: 1,
+	}
+	client.commandable.process = client.process
+	client.reloadIfDue()
+	return client, nil
+}
+
+// Closes the pool
+func (c *ClusterClient) Close() error {
+	c.cachemx.Lock()
+	defer c.cachemx.Unlock()
+
+	return c.reset()
+}
+
+// GetMasterAddrBySlot finds the current master address for a given hash slot
+func (c *ClusterClient) GetMasterAddrBySlot(hashSlot int) string {
```

Comment:
The use case for this is creating pipelines. I was thinking about adding an `OmniPipeline` which would use the code above to group commands by address, then execute in individual Pipelines against the actual master, but until then, this is the only way to access this functionality